content/ai/concepts/prompt-engineering/prompt-engineering.md

```diff
@@ -0,0 +1,101 @@
+
+When an LLM is given a body of text it typically breaks this text into individual units or "tokens." These tokens are the base unit of information that is being sent to the LLM. You can think of a word as a token. Additional information about tokens can be found in [OpenAI's help docs](https://help.openai.com/en/articles/4936856-what-are-tokens-and-how-to-count-them). It then processes these tokens in order to generate an answer. This computation comes at a cost, and is the reason why companies charge a certain amount of dollars per token, particularly when these tokens are being sent through some sort of API request. In these situations, having clear language can not only save money, but also directly affect the performance of the LLM.
+
+One example of using unclear language could be the following:
+
+```
+Explain the concept of for loops in Javascript. Keep your explaination short, and do not be too descriptive.
+```
+
+In this example, there is too much ambiguity for an LLM. For example, there is no guidance on what "short" means, and "not be too descriptive" could have two meanings. Either, do not use too many words, which we already said, or provide a description that is easy to understand. A better prompt would be the following.
+
+```
+Use 2 to 3 sentences to explain the concept of for loops in Javascript to a 5 year old.
+```
+
+### Define Your Purpose
+
+Another key to writing effective prompts is defining the purpose of the prompt. In other words, effectively defining the goal for the LLM to achieve. This can be done by specifying the tone, format and audience of the prompt.
+
+One area in which tone is important, is writing an email. For example, a different tone would be used when writing an email to a sibling, versus one to someone's boss. Therefore, when asking an LLM to write a draft of an email, it is important to make sure it understands what tone is preferred.
+
+In addition to specifying the tone, you can also specify the format of the LLM's response. In some situations it can be specified that the LLM use a bullet list, versus a series of paragraphs.
+
+Finally, by specifying a possible audience, an LLM can even be more precise when generating text. For instance, an LLM will most likely provide additional details when writing an email for a professional audience, versus a less technical audience.
+
+### Provide Context
+
+As mentioned earlier, LLMs are trained using vast amounts of data. It uses this data to effectively predicate and craft responses to a user's prompt. One way to indirectly control what data an LLM uses in its response, is by clearly defining the context of the prompt.
+
+A common use case in which context is important is when asking an LLM to debug an error message from code. This itself can be an entire area of study because there can be many types of errors. For example, when encountering a specific error associated with a line of code, a good prompt would be the following.
```

Comment:
The part about debugging code being an entire area of study should be removed from this sentence as it can be a bit distracting.